src/main/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentStepExecution.java

```diff
@@ -0,0 +1,195 @@
+package com.cloudbees.jenkins.plugins.sshagent;
+
+import com.cloudbees.jenkins.plugins.sshcredentials.SSHUserPrivateKey;
+import com.cloudbees.plugins.credentials.CredentialsProvider;
+import com.google.inject.Inject;
+import hudson.EnvVars;
+import hudson.Launcher;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.util.Secret;
+import jenkins.model.Jenkins;
+import org.apache.commons.lang.StringUtils;
+import org.jenkinsci.plugins.workflow.steps.*;
+
+import javax.annotation.CheckReturnValue;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.*;
+
+public class SSHAgentStepExecution extends AbstractStepExecutionImpl {
+
+    private static final long serialVersionUID = -1912198498615126332L;
+
+    @StepContextParameter
+    private transient TaskListener listener;
+
+    @StepContextParameter
+    private transient Run build;
+
+    @StepContextParameter
+    private transient Launcher launcher;
+
+    @Inject
+    private transient SSHAgentStep step;
+
+    /**
+     * The proxy for the real remote agent that is on the other side of the channel (as the agent needs to
+     * run on a remote machine)
+     */
+    private transient RemoteAgent agent = null;
+
+    @Override
+    public boolean start() throws Exception {
+        StepContext context = getContext();
+        String socket = initRemoteAgent();
+        context.newBodyInvoker().
+                withContext(EnvironmentExpander.merge(getContext().get(EnvironmentExpander.class), new ExpanderImpl(socket))).
+                withCallback(new Callback()).withDisplayName(null).start();
+        return false;
+    }
+
+    @Override
+    public void stop(Throwable cause) throws Exception {
+        if (agent != null) {
+            agent.stop();
+            listener.getLogger().println(Messages.SSHAgentBuildWrapper_Stopped());
+        }
+    }
+
+    @Override
+    public void onResume() {
+        super.onResume();
+        try {
+            StepContext context = getContext();
+            String socket = initRemoteAgent();
+            context.newBodyInvoker().
+                    withContext(EnvironmentExpander.merge(getContext().get(EnvironmentExpander.class), new ExpanderImpl(socket))).
+                    withCallback(new Callback()).withDisplayName(null).start();
```

Comment:
Actually `StepExecutionIterator` will probably not help. Not sure offhand how to look up the contextual `ExpanderImpl` from `onResume`’s `StepContext`. Calling `context.get(ExpanderImpl.class)` will not work since `EnvironmentExpander.merge` registers a proxy.